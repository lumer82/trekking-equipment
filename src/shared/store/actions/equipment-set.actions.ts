import { Action } from '@ngrx/store';

export const EquipmentSetActionTypes = {
  CREATE_NEW: '[EquipmentSet] Create new'
};

export class EquipmentSetCreateNewAction implements Action {
  readonly type: string = EquipmentSetActionTypes.CREATE_NEW;
}

export type EquipmentSetActions =
  EquipmentSetCreateNewAction;
