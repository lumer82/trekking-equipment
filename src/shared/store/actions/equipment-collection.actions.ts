import { Action } from '@ngrx/store';
import { EquipmentCollection } from '../reducer/equipment-collection.reducer';

export const EquipmentCollectionActionTypes = {
  ADD: '[EquipmentCollection] Add'
};

export class EquipmentCollectionAddAction implements Action {
  readonly type: string = EquipmentCollectionActionTypes.ADD;

  constructor(readonly payload: EquipmentCollection) {}
}

export type EquipmentCollectionActions =
  EquipmentCollectionAddAction;
