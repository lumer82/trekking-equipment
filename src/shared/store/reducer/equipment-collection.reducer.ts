import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import { Action } from '@ngrx/store';
import { EquipmentSetActions, EquipmentSetActionTypes } from '../actions/equipment-set.actions';
import { equipmentEntriesReducer, EquipmentEntriesState } from './equipment-entries.reducer';
import {
  EquipmentCollectionActions, EquipmentCollectionActionTypes,
  EquipmentCollectionAddAction
} from '../actions/equipment-collection.actions';
import { AddAction } from '../util/add-action';

export interface EquipmentCollection {
  id: string;
  name: string;
  entries: EquipmentEntriesState;
}

export type EquipmentCollectionsState = EntityState<EquipmentCollection>;

const adapter: EntityAdapter<EquipmentCollection> = createEntityAdapter<EquipmentCollection>();

const initialState: EquipmentCollectionsState = adapter.getInitialState();

export function equipmentCollectionsReducer(state: EquipmentCollectionsState = initialState,
                                            action: EquipmentSetActions | EquipmentCollectionActions): EquipmentCollectionsState {

  const addAction = (<AddAction>action).addAction;
  Object.values(state.entities)
    .filter(e => !addAction || (<AddAction>action).collection.id === e.id)
    .forEach(e => {
      const entries = equipmentEntriesReducer(e.entries, action);
      if (entries !== e.entries) {
        state = adapter.updateOne({id: e.id, changes: {...e, entries}}, state);
      }
    });

  switch (action.type) {
    case EquipmentSetActionTypes.CREATE_NEW:
      return initialState;
    case EquipmentCollectionActionTypes.ADD:
      return adapter.addOne((<EquipmentCollectionAddAction> action).payload, state);
    default:
      return state;
  }
}
