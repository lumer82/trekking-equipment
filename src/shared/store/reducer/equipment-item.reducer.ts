import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import { Action } from '@ngrx/store';
import { EquipmentSetActionTypes } from '../actions/equipment-set.actions';
import {EquipmentItemActionTypes, EquipmentItemAddAction} from '../actions/equipment-item.actions';

export interface EquipmentItem {
  id: string;
  name: string;
  price: number;
  weight: number;
}

export type EquipmentItemsState = EntityState<EquipmentItem>;

const adapter: EntityAdapter<EquipmentItem> = createEntityAdapter<EquipmentItem>();

const initialState: EquipmentItemsState = adapter.getInitialState();

export function equipmentItemsReducer(state: EquipmentItemsState = initialState, action: Action): EquipmentItemsState {
  switch (action.type) {
    case EquipmentSetActionTypes.CREATE_NEW:
      return initialState;
    case EquipmentItemActionTypes.ADD:
      return adapter.addOne((<EquipmentItemAddAction>action).payload, state);
    default:
      return state;
  }
}
