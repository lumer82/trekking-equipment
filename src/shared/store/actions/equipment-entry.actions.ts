import { AddAction } from '../util/add-action';
import { EquipmentCollection } from '../reducer/equipment-collection.reducer';
import { EquipmentEntry } from '../reducer/equipment-entries.reducer';

export const EquipmentEntryActionTypes = {
  ADD: '[EquipmentEntryAction] Add'
};

export class EquipmentEntryAddAction implements AddAction {
  readonly addAction: boolean = true;
  readonly type: string = EquipmentEntryActionTypes.ADD;

  constructor(public readonly collection: EquipmentCollection,
              public readonly payload: EquipmentEntry) {}
}

export type EquipmentEntryActions =
  EquipmentEntryAddAction;
