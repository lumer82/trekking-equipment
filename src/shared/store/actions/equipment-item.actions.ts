import {AddAction} from '../util/add-action';
import {EquipmentCollection} from '../reducer/equipment-collection.reducer';
import {EquipmentEntry} from '../reducer/equipment-entries.reducer';
import {EquipmentItem} from '../reducer/equipment-item.reducer';

export const EquipmentItemActionTypes = {
  ADD: '[EquipmentItem] Add'
};

export class EquipmentItemAddAction implements AddAction {
  readonly addAction: boolean = true;
  readonly type: string = EquipmentItemActionTypes.ADD;

  constructor(public readonly collection: EquipmentCollection,
              public readonly entry: EquipmentEntry,
              public readonly payload: EquipmentItem) {}
}
