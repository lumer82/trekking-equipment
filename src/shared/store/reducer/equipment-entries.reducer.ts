import {createEntityAdapter, EntityAdapter, EntityState} from '@ngrx/entity';
import {Action} from '@ngrx/store';
import {EquipmentSetActionTypes} from '../actions/equipment-set.actions';
import {equipmentItemsReducer, EquipmentItemsState} from './equipment-item.reducer';
import {AddAction} from '../util/add-action';
import {EquipmentEntryActionTypes, EquipmentEntryAddAction} from '../actions/equipment-entry.actions';

export interface EquipmentEntry {
  id: string;
  name: string;
  items: EquipmentItemsState;
}

export interface EquipmentEntriesState extends EntityState<EquipmentEntry> {
  selectedItemId: string | null;
}

const adapter: EntityAdapter<EquipmentEntry> = createEntityAdapter<EquipmentEntry>();

const initialState: EquipmentEntriesState = adapter.getInitialState({selectedItemId: undefined});

export function equipmentEntriesReducer(state: EquipmentEntriesState = initialState, action: Action): EquipmentEntriesState {
  const addAction = (<AddAction>action).addAction;
  Object.values(state.entities)
    .filter(e => !addAction || (<AddAction>action).entry.id === e.id)
    .forEach(e => {
      const items = equipmentItemsReducer(e.items, action);
      if (items !== e.items) {
        state = adapter.updateOne({id: e.id, changes: {...e, items}}, state);
      }
    });

  switch (action.type) {
    case EquipmentSetActionTypes.CREATE_NEW:
      return initialState;
    case EquipmentEntryActionTypes.ADD:
      return adapter.addOne((<EquipmentEntryAddAction>action).payload, state);
    default:
      return state;
  }
}
