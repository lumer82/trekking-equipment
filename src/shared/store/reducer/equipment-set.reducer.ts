import { equipmentCollectionsReducer, EquipmentCollectionsState } from './equipment-collection.reducer';
import { Action } from '@ngrx/store';
import { EquipmentSetActionTypes } from '../actions/equipment-set.actions';

export interface EquipmentSetState {
  _id?: string;
  name: string;
  collections: EquipmentCollectionsState;
}

const initialState: EquipmentSetState = {
  name: '',
  collections: undefined
}

export function equipmentSetReducer(state: EquipmentSetState = initialState, action: Action): EquipmentSetState {
  const collections = equipmentCollectionsReducer(state.collections, action);
  if (collections !== state.collections) {
    state = {...state, collections};
  }

  switch (action.type) {
    case EquipmentSetActionTypes.CREATE_NEW:
      return {...initialState};
    default:
      return state;
  }

}
