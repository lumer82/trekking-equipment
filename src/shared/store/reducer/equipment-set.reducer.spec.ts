import {equipmentSetReducer, EquipmentSetState} from './equipment-set.reducer';
import {EquipmentCollectionAddAction} from '../actions/equipment-collection.actions';
import {EquipmentCollection} from './equipment-collection.reducer';
import {EquipmentEntry} from './equipment-entries.reducer';
import {EquipmentItem} from './equipment-item.reducer';
import {EquipmentEntryAddAction} from '../actions/equipment-entry.actions';
import {EquipmentItemAddAction} from '../actions/equipment-item.actions';

describe('EquipmentSetReducer', () => {
  describe('add', () => {
    const collection: EquipmentCollection = {id: 'collection_1', name: 'collection_1', entries: undefined};
    const entry: EquipmentEntry = {id: 'entry_1', name: 'entry_1', items: undefined};
    const item: EquipmentItem = {id: 'item_1', name: 'item_1', weight: undefined, price: undefined};
    it('should add a new collection', () => {
      const actual = equipmentSetReducer(undefined, new EquipmentCollectionAddAction(collection));

      console.log('collection is', actual.collections);
      expect(actual.collections.entities[collection.id]).toBe(collection);
    });

    it('should add a new entry to given collection', () => {
      const state: EquipmentSetState = {
        name: 'set_1',
        collections: {
          ids: [collection.id],
          entities: {
            'collection_1': collection
          }
        }
      };

      const actual = equipmentSetReducer(state, new EquipmentEntryAddAction(collection, entry));
      expect(actual).toMatchSnapshot();
    })

    it('should add a new item to given collection and entry', () => {
      const state: EquipmentSetState = {
        name: 'set_1',
        collections: {
          ids: [collection.id],
          entities: {
            'collection_1':
              {
                ...collection, entries: {
                  selectedItemId: undefined,
                  ids: [entry.id],
                  entities: {
                    'entry_1': entry
                  }
                }
              }
          }
        }
      };

      const actual = equipmentSetReducer(state, new EquipmentItemAddAction(collection, entry, item));
      expect(actual).toMatchSnapshot();
    });
  });
});
