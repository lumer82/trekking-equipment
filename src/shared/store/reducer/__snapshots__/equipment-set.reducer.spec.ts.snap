// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EquipmentSetReducer add should add a new entry to given collection 1`] = `
Object {
  "collections": Object {
    "entities": Object {
      "collection_1": Object {
        "entries": Object {
          "entities": Object {
            "entry_1": Object {
              "id": "entry_1",
              "items": undefined,
              "name": "entry_1",
            },
          },
          "ids": Array [
            "entry_1",
          ],
          "selectedItemId": undefined,
        },
        "id": "collection_1",
        "name": "collection_1",
      },
    },
    "ids": Array [
      "collection_1",
    ],
  },
  "name": "set_1",
}
`;

exports[`EquipmentSetReducer add should add a new item to given collection and entry 1`] = `
Object {
  "collections": Object {
    "entities": Object {
      "collection_1": Object {
        "entries": Object {
          "entities": Object {
            "entry_1": Object {
              "id": "entry_1",
              "items": Object {
                "entities": Object {
                  "item_1": Object {
                    "id": "item_1",
                    "name": "item_1",
                    "price": undefined,
                    "weight": undefined,
                  },
                },
                "ids": Array [
                  "item_1",
                ],
              },
              "name": "entry_1",
            },
          },
          "ids": Array [
            "entry_1",
          ],
          "selectedItemId": undefined,
        },
        "id": "collection_1",
        "name": "collection_1",
      },
    },
    "ids": Array [
      "collection_1",
    ],
  },
  "name": "set_1",
}
`;
